### PR TITLE
Kat Perreira, Hotel,  Edges

### PR DESCRIPTION
**Pull request by kat-perreira**
# Hotel
Congratulations! You're submitting your assignment!

## Comprehension Questions
Question | Answer
:------------- | :-------------
Describe a design decision you had to make when working on this project. What options were you considering? What helped you make your final decision? | I initially had a room class. When I got to my booking manager, I realized it was unneeded because the room numbers could exist instead as an array.
Describe a concept that you gained more clarity on as you worked on this assignment. | On Saturday evening I realized the tests that I was writing wasn't actually testing my reservation maker method and I tried to fix too many things at once then lots of things broke. So I learned, again, to fix one thing at a time. Up until that point I had been doing test driven development and I probably wouldn't have ever noticed that it wasn't working the way I wanted if I hadn't been doing that. I liked doing TDD, it helped keep me organized and forced me to think thru the design before writing anything actual code.  
Describe a _nominal test_ that you wrote for this assignment. | A nominal test that I wrote was in the reservations which calculated total cost of the reservation
Describe an _edge case test_ that you wrote for this assignment. | An edge case test I wrote was to raise an error if the check_out date was entered before the check_in date in a reservation.
How do you feel you did in writing pseudocode first, then writing the tests and then the code? | I liked that it kept me organized and slowed me down to really think about what I wanted to do. Several times I started writing the test and couldn't figure out what the method I planned to write would return, and I think that was a good thing. It did cause me to spot the error which completely derailed me and i'm sad that I ran out of time but I guess that's better then typing along thinking everything is working when it isn't.

### CODE REVIEW

**Comment by tildeee** [comment]
# Hotel
## What We're Looking For

Feature | Feedback
---     | ---
**Baseline** |
Used git regularly	| x
Answer comprehension questions	| x
**Design** |
Each class is responsible for a single piece of the program | x
Classes are loosely coupled | x
**Wave 1** |
List rooms | x
Reserve a room for a given date range | doesn't work!
List reservations for a given date | 
Calculate reservation price | x
Invalid date range produces an error | x
Test coverage | not great
**Wave 2** |
View available rooms for a given date range | doesn't work
Reserving a room that is not available produces an error | 
Test coverage | 
**Wave 3** |
Create a block of rooms | 
Check if a block has rooms | 
Reserve a room from a block | 
Test coverage | 
**Fundamentals** | 
Names variables, classes and modules appropriately | there are some names of variables that have poor names that I think is tied in with the following point...
Understanding of variable scope - local vs instance | I'm noticing a huge misuse of local variables vs. instance variables-- let's talk soon
Can create complex logical structures utilizing variables | there are a couple of places of broken code here
Appropriately uses methods to break down tasks into smaller simpler tasks | the logic is pretty strange
Understands the differences between class and instance methods | there are instance methods that aren't declared correctly
Appropriately uses iterators and enumerables | the loops here rely on while loops in ways that are difficult to parse, or a broken times loop
Appropriately writes and utilizes classes | work in progress
Appropriately utilizes modules as a mechanism to encapsulate functionality | no modules in this project, which is okay
**Wrap Up** |
There is a refactors.txt file | x
The file provides a roadmap to future changes | I want to see refactors that aren't "fix this" or "implement this" but are more about how to restructure existing functionality
**Additional Feedback** | 

Hey Kat! I see some red flags in your code that I'd like to address.


I'll be making comments in the code itself, but the overall themes are:
1. **I'm seeing places where code is in the wrong place**-- wrong areas of files
1. I'm seeing many places where local variables and/or instance variables are being declared in ways that conflict and result in broken code
1. I'm seeing a place where the name of an instance variable is colliding with an `attr_accessor` and a method name, which is breaking your code
1. I'm seeing lack of clarity in loops
1. I'm seeing lack of clarity in some complex conditional logic


Overall, I think your code has a tremendous lack of nominal tests. I see your progress in tests to cover the edge cases, but I would rather have tests around the nominal cases first. I think there's a lot to simply establish with tests.

I know that you mentioned a couple of times that you broke the code at the last minute. However, beyond any code that was broken at the last minute, I'm seeing the large broad strokes of setting up the internals of your design and paving the way. I see the outline of some complex logic that doesn't seem to be going in the right direction, and I look for that setup closely, even if it was broken at the last minute.

Anyway, the important thing for us moving forward is that you feel like you have a list of observations I've made and you consider how you address the gaps of understanding that I'd like you go over soon. Please reach out to Dan or I if you have any questions

File: lib/booking_manager.rb [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+  def initialize
+    @start_date = start_date
+    @end_date = end_date
+    @reservations = []
+    @rooms = []
+    i = 0
+    20.times do |i|
+      @rooms << i + 1
+      i += 1
+    end
```
Comment (tildeee) [inline lib/booking_manager.rb]: In `times` loops, we get an iteration variable. In this case, on line 16, you make the iteration variable named `i`, and it ends up being the value that the `20.times` part makes it.

However, outside of the loop, you declare `i = 0`, and then inside the loop, you increment `i +=1` in line 18... Even though the times loop should take care of this, making lines 15 and 18 unnecessary.

This is an indicator of a misunderstanding about the times loop -- please review the times loop, and reach out if you want to go over it

File: lib/booking_manager.rb [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+
+  end
+
+  def find_available_room(start_date, end_date)
+    requested_dates = (start_date...end_date).to_a
+    room = 237
```
Comment (tildeee) [inline lib/booking_manager.rb]: Why are you saying `room = 237`? I understand you might just need a random number, but in programming, we should intentionally pick a placeholder value, such as 0 or `nil`

File: lib/booking_manager.rb [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+    while room_count < 20
+      if @reservations.empty?
+        room = 1
+        return room
+      end
+      @reservations.each_with_index do |reservation, index|
+        if reservation.room == (index + 1) && (reservation.reservation_dates & requested_dates).length == 0
+          room = index + 1
+          return room
+        elsif reservation.room == (index + 1)
```
Comment (tildeee) [inline lib/booking_manager.rb]: I don't understand why we need to check if this is the case. What does this represent?

File: lib/booking_manager.rb [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+          room_count += 1
```
Comment (tildeee) [inline lib/booking_manager.rb]: ? What is `room_count` for? Is it just for helping us iterate with the while loop `while room_count < 20`? Why does room_count only increment in this case? Why does it have to be in a while loop? Is there a clearer way to express this without a while loop?

File: lib/booking_manager.rb [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+        elsif reservation.room != (index + 1)
+          room = index + 1
+          return room
+        end
+      end
```
Comment (tildeee) [inline lib/booking_manager.rb:53]: Overall this `if ... elsif ... elsif` block feels pretty unclear to me about what's going on -- maybe making tests, helper methods, adding comments, or trying a different structure between `if ... elsif ... elsif` would help

File: lib/booking_manager.rb [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+
+def make_reservation(start_date, end_date)
+  start_date = @start_date
+  end_date = @end_date
+
+  room = find_available_room(start_date, end_date)
+  if @rooms.include? room
+    reservation = Reservation.new(room, start_date, end_date)
+    @reservations << reservation
+    binding.pry
```
Comment (tildeee) [inline lib/booking_manager.rb]: Please take out your `binding.pry`s before project submission!

File: lib/booking_manager.rb [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+    @make_reservation
```
Comment (tildeee) [inline lib/booking_manager.rb]: What is this line trying to do? This line, as-is, says "Invoke the value that is stored on the instance variable named `@make_reservation`," which is not defined anywhere else in the code... You have a method named `make_reservation`?

File: lib/booking_manager.rb [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+  end
+end
```
Comment (tildeee) [inline lib/booking_manager.rb:72]: This entire method `make_reservation` is outside of the class `BookingManager` ... literally -- outside of the `class BookingManager ... end`!

File: lib/booking_manager.rb [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+        end
+      end
+    end
+    raise StandardError.new("No room available")
+  end
+end
+
+def make_reservation(start_date, end_date)
+  start_date = @start_date
+  end_date = @end_date
```
Comment (tildeee) [inline lib/booking_manager.rb]: What's going on in these lines?

In line 48, you define a method named `make_reservation`, which takes in two pieces of data. The first parameter becomes the local variable named `start_date`, and the second parameter becomes the local variable named `end_date`.

In line 49, you say that the value of the local variable `start_date` is ACTUALLY reassigned to the value of the instance variable `@start_date` ... which, in most cases of your code running, is `nil`, because it's not defined (correctly) before this.

Similarly, in line 50, you say that the value of the local variable `end_date` is ACTUALLY reassigned to the value of the instance variable `@end_date`, or `nil`.

What were you trying to do?

For me, this says you aren't sure about how to use instance variables and local variables. Reach out to me or Dan if you want to go over it.

File: lib/booking_manager.rb [hunk 10]
```diff
@@ -0,0 +1,59 @@
...
+
+class BookingManager
+
+  attr_reader :rooms
+  attr_accessor :reservations, :start_date, :end_date, :make_reservation
+
+
+  def initialize
+    @start_date = start_date
+    @end_date = end_date
```
Comment (tildeee) [inline lib/booking_manager.rb:12]: What's happening in the above two lines?

In this class `BookingManager`, you are saying that you want an instance variable `@start_date` that an instance of BookingManager holds. However, you're saying on line 11 that the value of `@start_date` should be assigned to the value of the local variable `start_date`... which isn't defined anywhere before that! Therefore, `@start_date`'s value becomes `nil`

Same thing for `@end_date`

File: lib/reservation_manager.rb [hunk 11]
```diff
@@ -0,0 +1,7 @@
+
+class ReservationManager
```
Comment (tildeee) [inline lib/reservation_manager.rb]: Please delete the files you end up not needing before project submission!

File: spec/booking_manager_spec.rb [hunk 12]
```diff
@@ -0,0 +1,60 @@
...
+  end
+
+end
+
+
+describe '#find_available_rooms' do
+  # fully_booked: hotel filled from 2018, 1, 1 til 2018 1, 7
+  # - Same dates
+  it 'must raise standard error if two reservations made for the same date' do
+    expect{fully_booked.make_reservation(Date.new(2018, 1, 1), Date.new(2018, 1, 7))}.must_raise StandardError
```
Comment (tildeee) [inline spec/booking_manager_spec.rb]: You end up referring to the variable `fully_booked` in this line...

`fully_booked` is going to be undefined.

That's because you define `fully_booked` over on line 12, in a scope that doesn't apply to these tests at all.

These tests pass because this line ends up raising an Error indeed -- an error caused by `fully_booked.make_reservation` not existing.

Comment (tildeee) [reply spec/booking_manager_spec.rb]: Let me or Dan know if there are questions about why `fully_booked` is out of scope and undefined.

File: spec/reservation_manager_spec.rb [hunk 13]
```diff
@@ -0,0 +1 @@
+require_relative 'spec_helper.rb'
```
Comment (tildeee) [inline spec/reservation_manager_spec.rb]: Don't forget to remove files that you end up not needing!

File: spec/reservation_spec.rb [hunk 14]
```diff
@@ -0,0 +1,64 @@
...
+    it 'Calculates the amount of time that the reservation is active' do
+      #general reservation for 6 nights
+      expect (general_reservation.stay_duration).must_equal 6
+    end
+  end
+
+  describe "#total_charge" do
+    it 'Charges 200.00 a night for a reservation' do
+      # general reservation is for 7 days, 6 nights, 6 * 200 = 1200
+      expect(general_reservation.total_charge(6)).must_equal 1200
```
Comment (tildeee) [inline spec/reservation_spec.rb:59]: You end up calling `general_reservation.total_charge(6)` ... You are explicitly passing in `6` into the method `total_charge`. Why? Why does this test have to know that we have to pass in 6?

If `general_reservation` is an instance of `Reservation` that knows its own start date, end date, duration, etc, then why do we have to pass in `6`? Why can't it calculate everything internal to itself?

File: lib/group_reservations.rb [hunk 15]
```diff
@@ -0,0 +1,19 @@
+require_relative 'reservation.rb'
+require_relative 'booking_manager.rb'
+
+
+class GroupReservations < Reservation
```
Comment (tildeee) [inline lib/group_reservations.rb]: Don't forget to clean up your files!

File: lib/group_reservation.rb [hunk 16]
```diff
@@ -0,0 +1,19 @@
...
+class GroupReservation < Reservation
+  attr_reader :room
+  attr_accessor :check_in
+  attr_accessor :check_out
+  attr_reader :reservation_dates
+  attr_accessor :group_name
+
+  def initialize(room, check_in, check_out, group_name)
+    super(room, check_in, check_out)
+    @group_name = group_name
```
Comment (tildeee) [inline lib/group_reservation.rb:14]: Was there a requirement to save a group name?

File: lib/booking_manager.rb [hunk 17]
```diff
@@ -0,0 +1,59 @@
+require_relative 'reservation.rb'
+require 'pry'
+
+class BookingManager
+
+  attr_reader :rooms
+  attr_accessor :reservations, :start_date, :end_date, :make_reservation
```
Comment (tildeee) [inline lib/booking_manager.rb:7]: We use `attr_`... methods at the top of classes in order to connect _instance variables_ to levels of access-- reading, writing, or both.

`BookingManager` doesn't have an instance variable named `@make_reservation`. It has a method named `make_reservation`, but methods don't belong here.

What were you trying to do?

Let me or Dan know if you need to sit down with someone to work through this misunderstanding.
